install/0000_90_cluster-version-operator_02_servicemonitor.yaml

```diff
@@ -16,3 +16,49 @@ spec:
   selector:
     matchLabels:
       k8s-app: cluster-version-operator
+---
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  labels:
+    k8s-app: cluster-version-operator
+  name: cluster-version-operator
+  namespace: openshift-cluster-version
+spec:
+  groups:
+  - name: cluster-version
+    rules:
+    - alert: ClusterVersionOperatorDown
+      annotations:
+        message: Cluster version operator has disappeared from Prometheus target discovery. Operator may be down or disabled, cluster will not be kept up to date and upgrades will not be possible.
+      expr: |
+        absent(up{job="cluster-version-operator"} == 1)
+      for: 15m
+      labels:
+        severity: critical
+  - name: cluster-operators
+    rules:
+    - alert: ClusterOperatorDown
+      annotations:
+        message: Cluster operator {{ "{{ $labels.name }}" }} has not been available for 10 mins. Operator may be down or disabled, cluster will not be kept up to date and upgrades will not be possible.
+      expr: |
+        cluster_operator_up{job="cluster-version-operator"} == 0
+      for: 10m
+      labels:
+        severity: critical
+    - alert: ClusterOperatorDegraded
+      annotations:
+        message: Cluster operator {{ "{{ $labels.name }}" }} has been degraded for 10 mins. Operator is degraded because {{ "{{ $labels.reason }}" }} and cluster upgrades will be unstable.
+      expr: |
+        cluster_operator_conditions{job="cluster-version-operator", condition="Degraded"} == 1
+      for: 10m
+      labels:
+        severity: critical
+    - alert: ClusterOperatorFlapping
+      annotations:
+        message: Cluster operator {{ "{{ $labels.name }}" }} up status is changing often. This might cause upgrades to be unstable.
+      expr: |
+        changes(cluster_operator_up{job="cluster-version-operator"}[2m]) > 2
```

Comment:
2m range selector doesn't seem to cover sufficient failure cases, also this behavior seems highly specific to each operator. Some operators have a very short reconciliation loop, some have a very long one. I don't think we're going to be able to find any good general purpose value for this.

Reply:
The clusteroperators should not be flapping between conditions. It has nothing to do with reconcile times.

> https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#what-should-an-operator-report-with-clusteroperator-custom-resource

for example,
> An operator reports Progressing when it is rolling out new code, propagating config changes, or otherwise moving from one steady state to another. It should not report progressing when it is reconciling a previously known state

that makes it pretty clear that reconciles of current behavior making clusteroperator progressing is not recommended.

Reply:
It's unclear that that's what you want to test for. Also this sounds like it rather belongs in an e2e test than an alert.

Reply:
We need to know whether it is happening in the wild.  I think we’ve shown conclusively in 4.1 with telemetry that our internal testing of operators is insuffficient to catch what is happening in the wild.

I want to prove these things aren’t happening in the wild (and we have evidence it is) and maybe we can remove the alert later if it isn’t firing in the wild.  But because this is a critical control loop extra alerts are a good thing.

Reply:
To prove this isn't happening in the wild severity warning is sufficient. I am ok with adding this to prove that it's not happening.